src/blockencodings.h

```diff
@@ -12,58 +12,70 @@
+};
 
-    template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action) {
-        READWRITE(tx); //TODO: Compress tx encoding
+template<typename I>
+struct Uint48Wrapper {
+private:
+    I& m_int;
+public:
+    Uint48Wrapper(I& i) : m_int(i) {}
+    template <typename Stream> void Serialize(Stream& s) const
+    {
+        uint32_t lsb = m_int & 0xffffffff;
+        uint16_t msb = (m_int >> 32) & 0xffff;
+        s << lsb << msb;
+    }
+    template <typename Stream> void Unserialize(Stream& s)
+    {
+        uint32_t lsb;
+        uint16_t msb;
+        s >> lsb >> msb;
+        m_int = (uint64_t(msb) << 32) | uint64_t(lsb);
     }
 };
 
+template<typename T>
+static inline TransactionCompressWrapper<T> TransactionCompressor(T& tx) { return TransactionCompressWrapper<T>(tx); }
+
 class BlockTransactionsRequest {
 public:
     // A BlockTransactionsRequest message
     uint256 blockhash;
     std::vector<uint16_t> indexes;
 
-    ADD_SERIALIZE_METHODS;
-
-    template <typename Stream, typename Operation>
-    inline void SerializationOp(Stream& s, Operation ser_action) {
-        READWRITE(blockhash);
+    template <typename Stream>
+    void Serialize(Stream& s) const
+    {
+        s << blockhash;
         uint64_t indexes_size = (uint64_t)indexes.size();
-        READWRITE(COMPACTSIZE(indexes_size));
-        if (ser_action.ForRead()) {
-            size_t i = 0;
-            while (indexes.size() < indexes_size) {
-                indexes.resize(std::min((uint64_t)(1000 + indexes.size()), indexes_size));
-                for (; i < indexes.size(); i++) {
-                    uint64_t index = 0;
-                    READWRITE(COMPACTSIZE(index));
-                    if (index > std::numeric_limits<uint16_t>::max())
-                        throw std::ios_base::failure("index overflowed 16 bits");
-                    indexes[i] = index;
-                }
-            }
-
-            uint16_t offset = 0;
-            for (size_t j = 0; j < indexes.size(); j++) {
-                if (uint64_t(indexes[j]) + uint64_t(offset) > std::numeric_limits<uint16_t>::max())
-                    throw std::ios_base::failure("indexes overflowed 16 bits");
-                indexes[j] = indexes[j] + offset;
-                offset = indexes[j] + 1;
-            }
-        } else {
-            for (size_t i = 0; i < indexes.size(); i++) {
-                uint64_t index = indexes[i] - (i == 0 ? 0 : (indexes[i - 1] + 1));
-                READWRITE(COMPACTSIZE(index));
-            }
+        s << CompactSize(indexes_size);
+        for (size_t i = 0; i < indexes.size(); i++) {
+            const uint64_t index = indexes[i] - (i == 0 ? 0 : (indexes[i - 1] + 1));
+            s << CompactSize(index);
+        }
+    }
+
+    template <typename Stream>
+    void Unserialize(Stream& s)
+    {
+        std::vector<uint64_t> tmp;
+        s >> blockhash >> VectorApply<CompactSizeWrapper>(tmp);
```

Comment:
For ease of implementation, deserialization first happens into a uint64_t, and is then converted. This means a temporary is created and allocated, which is an overhead that the old implementation didn't have.

If this is considered too much, an alternative is to lose the overflow checking, and instead use `s >> blockhash >> VectorApply<CompactSizeWrapper>(indexes)` directly, which will work, but compute results modulo 2^16.